### PR TITLE
feat: Add availability zones to `avm/res/network/public-ip-prefix`

### LINKED ISSUE #1487: [AVM Module Issue]: Add the zones parameter to the `avm/res/network/public-ip-prefix` module
### Check for previous/existing GitHub issues

- [X] I have checked for previous/existing GitHub issues

### Issue Type?

Feature Request

### Module Name

avm/res/network/public-ip-prefix

### (Optional) Module Name if not listed above

_No response_

### (Optional) Module Version

_No response_

### Description

Please add the "zones" parameter. This will allow additional scenarios like, for instance, the use of the public-ip-prefix in the same AZ of the nat-gateway.

### (Optional) Correlation Id

_No response_

### PR DESCRIPTION
**Pull request by krbar**
## Description

Adds `zones` parameter to the module.

Resolves #1487 

## Pipeline Reference

| Pipeline |
| -------- |
|   [avm.res.network.public-ip-prefix](https://github.com/krbar/bicep-registry-modules/actions/workflows/avm.res.network.public-ip-prefix.yml)  |

## Type of Change

- [ ] Update to CI Environment or utlities (Non-module effecting changes)
- [x] Azure Verified Module updates:
  - [ ] Bugfix containing backwards compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [x] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

### CODE REVIEW

File: avm/res/network/public-ip-prefix/main.bicep [hunk 1]
```diff
@@ -26,6 +26,13 @@ param tags object?
 @description('Optional. The custom IP address prefix that this prefix is associated with. A custom IP address prefix is a contiguous range of IP addresses owned by an external customer and provisioned into a subscription. When a custom IP prefix is in Provisioned, Commissioning, or Commissioned state, a linked public IP prefix can be created. Either as a subset of the custom IP prefix range or the entire range.')
 param customIPPrefix object = {}
 
+@description('Optional. A list of availability zones denoting the IP allocated for the resource needs to come from.')
```
Comment (AlexanderSehr) [inline avm/res/network/public-ip-prefix/main.bicep:29]: Please update as per [BCPFR5 - Category: Inputs - Availability Zones Implementation](https://azure.github.io/Azure-Verified-Modules/specs/bicep/#id-bcpfr5---category-inputs---availability-zones-implementation).

To ease usuability we recommend to use 'primitive types' where possible. There is also a [PR](https://github.com/Azure/Azure-Verified-Modules/pull/803) open to re-enforce this.

Comment (krbar) [reply avm/res/network/public-ip-prefix/main.bicep:29]: Has this been successfully implemented yet?

In this case following implementation crashes the Set-AVMmodule:

```bicep
@description('Optional. A list of availability zones denoting the IP allocated for the resource needs to come from.')
@allowed([
  1
  2
  3
])
param zones int[] = [
  1
  2
  3
]

resource publicIpPrefix 'Microsoft.Network/publicIPPrefixes@2023-09-01' = {
  name: name
  location: location
  tags: tags
  sku: {
    name: 'Standard'
  }
  zones: [for zone in zones: string(zone)]
  properties: {
    customIPPrefix: !empty(customIPPrefix) ? customIPPrefix : null
    publicIPAddressVersion: 'IPv4'
    prefixLength: prefixLength
  }
}
```

But it works when I use a variable, like this:

```bicep
@description('Optional. A list of availability zones denoting the IP allocated for the resource needs to come from.')
@allowed([
  1
  2
  3
])
param zones int[] = [
  1
  2
  3
]

var zoneStrings = [for zone in zones: string(zone)]

resource publicIpPrefix 'Microsoft.Network/publicIPPrefixes@2023-09-01' = {
  name: name
  location: location
  tags: tags
  sku: {
    name: 'Standard'
  }
  zones: zoneStrings
  properties: {
    customIPPrefix: !empty(customIPPrefix) ? customIPPrefix : null
    publicIPAddressVersion: 'IPv4'
    prefixLength: prefixLength
  }
}
```

Comment (AlexanderSehr) [reply avm/res/network/public-ip-prefix/main.bicep:29]: Just ran some test. From a template perspective it looks fine and doesn't show an error, yet when I try to deploy Bicep get's stuck upon depoyment. It just doesn't do anything. Looks to me like a Bicep bug for which an issue should probably be created with the same content that you posted above (working vs. non-working) example.

That being said, if the variable unblocks you, let's add it. But we should then also update the spec so others are not running into the same issue. Could you do that once you finished this PR and we're sure the suggestion works?

File: avm/res/network/public-ip-prefix/main.bicep [hunk 2]
```diff
@@ -47,6 +59,8 @@ var builtInRoleNames = {
   )
 }
 
+var zoneStrings = [for zone in zones: string(zone)]
```
Comment (AlexanderSehr) [inline avm/res/network/public-ip-prefix/main.bicep:62]:
```suggestion
var formattedZones = [for zone in zones: string(zone)]
```

Comment (AlexanderSehr) [reply avm/res/network/public-ip-prefix/main.bicep:62]: Just a bit more generic ;)

File: avm/res/network/public-ip-prefix/main.bicep (hunk 1 above)
Comment (AlexanderSehr) [reply avm/res/network/public-ip-prefix/main.bicep:29]: AVM Spec update: https://github.com/Azure/Azure-Verified-Modules/pull/813

File: avm/res/network/public-ip-prefix/main.bicep [hunk 3]
```diff
@@ -73,6 +87,7 @@ resource publicIpPrefix 'Microsoft.Network/publicIPPrefixes@2023-09-01' = {
   sku: {
     name: 'Standard'
   }
+  zones: zoneStrings
```
[the remark this answers is not in the document]
Comment (krbar) [reply avm/res/network/public-ip-prefix/main.bicep]: @AlexanderSehr Updated and re-tested

File: avm/res/network/public-ip-prefix/main.bicep (hunk 1 above)
Comment (AlexanderSehr) [reply avm/res/network/public-ip-prefix/main.bicep:29]: @krbar I'm currently testing a [suggestion](https://github.com/Azure/bicep/issues/9668#issuecomment-2035405301) made by @jeskew via Pull Request #1521

If it works as hoped, we should apply the implementation here too and update the spec again

Comment (AlexanderSehr) [reply avm/res/network/public-ip-prefix/main.bicep:29]: Alright @krbar, this `map(zones, zone => string(zone))` does the trick as it does not result in a copy loop, but uses the new lambda functions.

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**